planner/core/optimizer_test.go

```diff
@@ -59,3 +59,53 @@ func (t *testPlannerFunctionSuite) TestMPPDecimalConvert(c *C) {
 	testDecimalConvert(20, 40, 0, 60, false, true, 20, 65, c)
 	testDecimalConvert(0, 40, 0, 60, false, false, 0, 60, c)
 }
+
+func testJoinKeyTypeConvert(leftType, rightType, retType *types.FieldType, lConvert, rConvert bool, c *C) {
+	cType, lCon, rCon := negotiateCommonType(leftType, rightType)
+	c.Assert(cType.Tp, Equals, retType.Tp)
+	c.Assert(cType.Flen, Equals, retType.Flen)
+	c.Assert(cType.Decimal, Equals, retType.Decimal)
+	c.Assert(cType.Flag, Equals, retType.Flag)
+	c.Assert(lConvert, Equals, lCon)
+	c.Assert(rConvert, Equals, rCon)
+}
+
+func (t *testPlannerFunctionSuite) TestMPPJoinKeyTypeConvert(c *C) {
+	tinyIntType := &types.FieldType{
+		Tp: mysql.TypeTiny,
+	}
+	tinyIntType.Flen, tinyIntType.Decimal = mysql.GetDefaultFieldLengthAndDecimal(mysql.TypeTiny)
+
+	unsignedTinyIntType := &types.FieldType{
+		Tp: mysql.TypeTiny,
+	}
+	unsignedTinyIntType.Flen, tinyIntType.Decimal = mysql.GetDefaultFieldLengthAndDecimal(mysql.TypeTiny)
+	unsignedTinyIntType.Flag = mysql.UnsignedFlag
+
+	bigIntType := &types.FieldType{
+		Tp: mysql.TypeLonglong,
+	}
+	bigIntType.Flen, tinyIntType.Decimal = mysql.GetDefaultFieldLengthAndDecimal(mysql.TypeLonglong)
+
+	unsignedBigIntType := &types.FieldType{
+		Tp: mysql.TypeLonglong,
+	}
+	unsignedBigIntType.Flen, tinyIntType.Decimal = mysql.GetDefaultFieldLengthAndDecimal(mysql.TypeLonglong)
+	unsignedBigIntType.Flag = mysql.UnsignedFlag
+
+	decimalType := &types.FieldType{
+		Tp:      mysql.TypeNewDecimal,
+		Flen:    20,
+		Decimal: 0,
+	}
+
+	testJoinKeyTypeConvert(tinyIntType, tinyIntType, tinyIntType, false, false, c)
```

Comment:
what about int/small int/big int vs decimal?

Reply:
TiDB guarantees that two type are in the same catalog, so do not need to test integer vs decimal